### PR TITLE
prov/efa: Fix efa_conn_alloc error path deadlock

### PR DESCRIPTION
**Pull request by brianz98**
When `efa_av_reverse_av_add` fails in `efa_conn_alloc`, the cleanup path acquires the SRX lock if `insert_implicit_av` is true. However, the only caller with `insert_implicit_av=true` is the CQ read path, which already holds the SRX lock. The SRX lock is a non-recursive mutex, so this double-lock would deadlock.

In fact, since the explicit path (`efa_av_insert`) also has the SRX lock, the locks in this cleanup path can be removed.

Unit tests to be added in the upcoming gtest framework.

### CODE REVIEW

**Comment by jiaxiyan** [comment]
The unit test should be in the same PR so I think this should come after your gtest framework

File: prov/efa/src/efa_conn.c [hunk 1]
```diff
@@ -303,13 +303,9 @@ struct efa_conn *efa_conn_alloc(struct efa_av *av, struct efa_ep_addr *raw_addr,
 	err = efa_av_reverse_av_add(av, cur_reverse_av, prv_reverse_av, conn);
 	if (err) {
 		if (av->domain->info_type == EFA_INFO_RDM) {
-			/* insert_implicit_av is only true for the CQ read path
-			 * which already has the SRX lock */
-			if (insert_implicit_av)
-				ofi_genlock_lock(&av->domain->srx_lock);
+			/* Both the CQ read path (implicit) and the explicit
```
Comment (jiaxiyan) [inline prov/efa/src/efa_conn.c]: what about the efa_rdm_cntr_progress path? It doesn't hold the srx lock

Comment (brianz98) [reply prov/efa/src/efa_conn.c]: https://github.com/ofiwg/libfabric/blob/ca57c5068f4e00682b96025daa4fafc789e996a8/prov/efa/src/efa_av.c#L492

this assert seems to suggest that the efa_rdm_cntr_progress path should also hold the srx lock?

Comment (jiaxiyan) [reply prov/efa/src/efa_conn.c]: nvm it is held in efa_cntr_read

**Comment by shijin-aws** [comment]
why having 2 commits operate on the same code lines?

File: prov/efa/test/gtest/efa_gtest_common_device.cc [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+void EfaMockDevice::set(EfaMockDevice *instance)
+{
+	g_mock_device = instance;
+}
+
+extern "C" {
+
+struct ibv_ah *__wrap_ibv_create_ah(struct ibv_pd *pd, struct ibv_ah_attr *attr)
```
Comment (sunkuamzn) [inline prov/efa/test/gtest/efa_gtest_common_device.cc]: Can we have a macro for this? Copy pasting these few lines for every mocked function will get tedious.

File: prov/efa/test/gtest/efa_gtest_common_device.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+#ifndef EFA_GTEST_COMMON_DEVICE_H
+#define EFA_GTEST_COMMON_DEVICE_H
+
+#include <gmock/gmock.h>
+#include <infiniband/verbs.h>
+#include <infiniband/efadv.h>
+
+class EfaMockDevice {
+public:
+	MOCK_METHOD(struct ibv_ah *, ibv_create_ah,
```
Comment (sunkuamzn) [inline prov/efa/test/gtest/efa_gtest_common_device.h]: I'm not sure if it is possible but if we can have macros handle the generation of the `MOCK_METHOD` element and the `__real*` prototypes, that would cut down so much boilerplate

File: prov/efa/test/gtest/efa_gtest_common_helpers.c [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+{
+	g_reverse_av_add_err = err;
+}
+
+int __real_efa_av_reverse_av_add(struct efa_av *av,
+				 struct efa_cur_reverse_av **cur_reverse_av,
+				 struct efa_prv_reverse_av **prv_reverse_av,
+				 struct efa_conn *conn);
+
+int __wrap_efa_av_reverse_av_add(struct efa_av *av,
```
Comment (sunkuamzn) [inline prov/efa/test/gtest/efa_gtest_common_helpers.c]: This looks like cmocka style mocking. Can we not set the mocked return value in the test?

File: prov/efa/test/gtest/efa_gtest_conn.cc [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+	void SetUp() override
+	{
+		memset(&resource, 0, sizeof(resource));
+	}
+
+	void TearDown() override
+	{
+		// Make sure to uninstall the mock
+		efa_test_set_reverse_av_add_err(0);
```
Comment (sunkuamzn) [inline prov/efa/test/gtest/efa_gtest_conn.cc]: Same here. cmocka style mocking.

File: prov/efa/src/efa_conn.c (hunk 1 above)
Comment (sunkuamzn) [inline prov/efa/src/efa_conn.c]: nit: no need for this comment. You could instead add `assert(ofi_genlock_held(&av->domain->srx_lock));`

File: prov/efa/test/gtest/efa_gtest_common_device.cc (hunk 2 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_device.cc]: Addressed in 86f9ff5

File: prov/efa/test/gtest/efa_gtest_common_device.h (hunk 3 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_device.h]: Addressed in 86f9ff5

File: prov/efa/test/gtest/efa_gtest_conn.cc (hunk 5 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_conn.cc]: Addressed in [5375808](https://github.com/ofiwg/libfabric/pull/12322/commits/537580822ae56bfb1f419afb523162b699d4b489)

File: prov/efa/test/gtest/efa_gtest_common_helpers.c (hunk 4 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_helpers.c]: Addressed in [5375808](https://github.com/ofiwg/libfabric/pull/12322/commits/537580822ae56bfb1f419afb523162b699d4b489)

**Comment by sunkuamzn** [comment]
Looks great, thanks!

2 minor comments

1. Can you add `using testing::_` and `using testing::Return` so that we don't have to keep typing "testing::" everywhere?

2. Can you squash and reorganize your commits so that the macro definitions are in the second commit and all mocked functions use the macros from the get go? It keeps the git history cleaner.

**Comment by brianz98** [comment]
> Looks great, thanks!
> 
> 2 minor comments
> 
> 1. Can you add `using testing::_` and `using testing::Return` so that we don't have to keep typing "testing::" everywhere?
> 2. Can you squash and reorganize your commits so that the macro definitions are in the second commit and all mocked functions use the macros from the get go? It keeps the git history cleaner.

Done and done! I've also added your nit for asserting the lock is held.

**Review by sunkuamzn — DISMISSED** [review]

File: prov/efa/test/gtest/efa_gtest_common_mocks.h [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+#define EFA_MOCK_PARAMS_efa_av_reverse_av_add                          \
+	struct efa_av *av, struct efa_cur_reverse_av **cur_reverse_av, \
+		struct efa_prv_reverse_av **prv_reverse_av,            \
+		struct efa_conn *conn
+#define EFA_MOCK_ARGS_efa_av_reverse_av_add \
+	av, cur_reverse_av, prv_reverse_av, conn
+
+/* --- Function list --- */
+
+#define EFA_MOCK_FUNCTIONS(X)             \
```
Comment (a-szegel) [inline prov/efa/test/gtest/efa_gtest_common_mocks.h:49]: Will future tests break if people add more mocks?

File: prov/efa/test/gtest/efa_gtest_common_helpers.h [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ * Skip the duplicates when the real EFA headers are already included
+ * (e.g. from the C implementation file) to avoid redefinition.
+ */
+#ifndef EFA_AV_H
+
+#ifndef EFA_GID_LEN
+#define EFA_GID_LEN 16
+#endif
+
+struct efa_ep_addr {
```
Comment (a-szegel) [inline prov/efa/test/gtest/efa_gtest_common_helpers.h]: Why do you need these here, can't you include them from somewhere else?

File: prov/efa/test/gtest/efa_gtest_conn.cc [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+TEST_F(EfaConnTest, alloc_reverse_av_add_failure_rdm_cleanup)
+{
+	fi_addr_t addr;
+	static struct ibv_ah dummy_ibv_ah;
+
+	efa_test_resource_construct(&resource, FI_EP_RDM, EFA_FABRIC_NAME);
+	ASSERT_NE(resource.ep, nullptr);
+
+	EfaMockDevice::set(&mock_device);
+	EXPECT_CALL(mock_device, ibv_create_ah(_, _))
```
Comment (a-szegel) [inline prov/efa/test/gtest/efa_gtest_conn.cc]: Do we want to make these tests stricter? None of them test parameters going in, and we have WillRepeately(Return(0))... which means always pass.

File: prov/efa/test/gtest/efa_gtest_common_mocks.h (hunk 6 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_mocks.h:49]: I'm assuming you're talking about mocked functions being invoked unintentionally, when real functions need to be called. This isn't an issue in this PR, but there are two solutions:

1. Define multiple mock classes analogous to `EfaMockDevice` today, to isolate the mocks, and only install the mocks you need by `EfaMockDevice::set(&mock_device);`
2. Set up the correct expectation to route all calls to a specific wrapped function to the real function, like
```
ON_CALL(*this, ibv_create_ah(_,:_))
        .WillByDefault(Invoke(__real_ibv_create_ah));
```
A mixture of both strategy can be used, and I expect that point to be reached shortly as we start to mock more functions with converging call paths.

File: prov/efa/test/gtest/efa_gtest_common_resource.cc [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	if (!fabric_name || !strcasecmp(fabric_name, EFA_DIRECT_FABRIC_NAME))
+		hints->mode |= FI_CONTEXT2;
+
+	if (ep_type == FI_EP_DGRAM)
+		hints->mode |= FI_MSG_PREFIX | FI_CONTEXT2;
+
+	return hints;
+}
+
+void efa_test_resource_construct(struct efa_resource *resource,
```
Comment (a-szegel) [inline prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: Do we want to make this more configurable off the bat?  We should pass in hints.

File: prov/efa/test/gtest/README.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+* `efa_gtest_common_helpers.{c,h}`: C-linkage helpers. EFA internal headers (`efa.h`, `efa_av.h`, ...) transitively pull in `unix/osd.h`, which uses C `_Complex` types that don't compile under C++. Anything a test needs from EFA internals is wrapped by a C function here and declared `extern "C"` in the header.
+* `efa_gtest_{component}.cc`: The tests themselves, one file per component (e.g. `efa_gtest_conn.cc`).
+
+## What Should be Tested
+1. We are biased toward testing public, stable functions, e.g. we prefer testing `fi_*` to `efa_*` - the former more closely resembles real user behavior.
+1. We make a conscious trade-off to test larger rather than smaller units. A large test unit consists of more state variables, and testing it gives us more confidence in the overall system.
+2. We are biased toward testing edge cases over "happy cases", especially if the code path under test cannot be covered by integration tests.
+
+## How to write
+1. Read the [GoogleTest documentation](https://google.github.io/googletest/).
```
Comment (a-szegel) [inline prov/efa/test/gtest/README.md]: Can you add links to primer and cookbook:
https://google.github.io/googletest/primer.html
https://google.github.io/googletest/gmock_cook_book.html

File: prov/efa/test/gtest/efa_gtest_common_mocks.h (hunk 6 above)
Comment (a-szegel) [reply prov/efa/test/gtest/efa_gtest_common_mocks.h:49]: Can you add this to READEME?  

What to do if you add a mock and break a bunch of tests?

**Review by jiaxiyan — COMMENTED** [review]
Commits should be atomic based on what they do. gtest setup in the first commit, bugfix with its unit tests in the second commit.

File: prov/efa/test/gtest/efa_gtest_common_helpers.h (hunk 7 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_helpers.h]: the .c file already includes `efa_av.h` and gets `efa_ep_addr` from there, but since the test .cc file [needs](https://github.com/brianz98/libfabric/blob/58146824fae5a58b2a2c8906b6e6232b97dcc080/prov/efa/test/gtest/efa_gtest_conn.cc#L67) `efa_ep_addr`, we can't include `efa_av.h` in the .h file, otherwise a `_Complex` definition will be pulled in, way upstream (in `unix/osd.h`), which will throw the C++ compiler. AFAICT this is just one of those unfortunate C/C++ interop issues we will have to manage (and hopefully minimize).

A permanent fix to this is to change the efa provider code to minimize exposure to 'unix/osd.h`, as Claude suggested:

> The cleanest fix: extract EFA_GID_LEN and struct efa_ep_addr into their own tiny header (e.g., prov/efa/src/efa_ep_addr.h) that doesn't transitively pull in _Complex. Then both efa_av.h and the gtest header include it directly — single source of truth, no duplication. The reason this works: the _Complex poison comes from unix/osd.h via efa.h → ofi.h → ofi_osd.h → unix/osd.h. But struct efa_ep_addr itself only needs <stdint.h>. A leaf header with just that struct and the define has no problematic transitive includes.

File: prov/efa/test/gtest/efa_gtest_common_resource.cc (hunk 9 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: Since it's hard to predict what will be needed in future tests (also potentially confusing in the context of this PR, i.e., adding unused args), I'm inclined to keep it simple for this PR and expand as we go. Is there something that stands out to you as being insufficient here?

Comment (a-szegel) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: I think we should pass in hints.

File: prov/efa/test/gtest/efa_gtest_common_helpers.h (hunk 7 above)
Comment (a-szegel) [reply prov/efa/test/gtest/efa_gtest_common_helpers.h]: Can you do a pre-declare? and drop the EFA_GID LEN?

File: prov/efa/test/gtest/efa_gtest_common_resource.cc (hunk 9 above)
Comment (jiaxiyan) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: Why do we copy these functions from cmocka instead of sharing them?

File: prov/efa/test/gtest/README.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
+# GoolgeTest-based EFA unit tests
```
Comment (jiaxiyan) [inline prov/efa/test/gtest/README.md]: nit: Goolge

File: prov/efa/test/gtest/efa_gtest_common_resource.cc (hunk 9 above)
Comment (a-szegel) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: We shouldn't add any dependencies between our google-test framework and our cmocka framework.

File: prov/efa/test/gtest/efa_gtest_common_helpers.h (hunk 7 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_helpers.h]: good point! I'll forward declare just the opaque struct and get rid of all this other stuff. This is already done in efa_gtest_common_device.h...

File: prov/efa/test/gtest/efa_gtest_common_resource.cc (hunk 9 above)
Comment (jiaxiyan) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: That is not what I asked

File: prov/efa/test/gtest/efa_gtest_common_mocks.h (hunk 6 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_mocks.h:49]: Will add to README. If that happens I think the two solutions above still apply: the test-writer either 1. define a new mock class and put his new mock and only the mocks they require into that class or 2. modify all broken tests to fall through to the __real_* functions. Again a combination of two should be used to cut down the number of necessary edits.

File: prov/efa/test/gtest/efa_gtest_common_resource.cc (hunk 9 above)
Comment (jiaxiyan) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: Sharing common functions does not add dependency between two frameworks

Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: Are you suggesting to have some common functions shared between CMocka and gTest? I copied the setup logic over because I'm anticipating changes to this file but not to the (currently largely similar or identical) CMocka file.

Comment (a-szegel) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: The dependency between the shared common functions is that we can't modify them in the gtest framework without risking breaking the cmocka framework.

Comment (jiaxiyan) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: These two functions look identical to efa_unit_test_resource_construct and efa_unit_test_resource_destruct. What changes are you anticipating?

Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: they're not identical today, the Cmocka versions hold a bit more logic (e.g., https://github.com/brianz98/libfabric/blob/58146824fae5a58b2a2c8906b6e6232b97dcc080/prov/efa/test/efa_unit_test_common.c#L274), and also use cmocka-provided assertion macros like `assert_int_equal`.

File: prov/efa/test/gtest/efa_gtest_conn.cc (hunk 8 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_conn.cc]: This test tests one thing, which is the efa_conn_alloc error path cleans up properly without crash or leak (this specific one tests the implicit path, the next one tests the explicit one). 

Having additional expectations might make these tests fail unexpectedly when the internal call paths change, and these spurious failures won't reveal whether a new bug is introduced to the the cleanup path.

As for the `WillRepeately(Return(0))`, we need these to always succeed to get to (and exit) the cleanup path.

File: prov/efa/test/gtest/efa_gtest_common_resource.cc (hunk 9 above)
Comment (jiaxiyan) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: That doesn't prevent you from putting shared logic into a common function. You can just use assert from plain C.

Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: I understand your concern about code duplication, but I've looked at the two functions and my opinion is that we will forego some (minor) nice features afforded to us by C++/gTest:

1. The failure control flow in CMocka and gTest are different: cmocka uses `goto`, gTest return through `ASSERT_EQ`. A shared function can't use either framework's assertions, so you'd need manual error-code checking plus a wrapper in each framework, adding a layer without reducing code.

2. Assertion semantics differ: gTest's `ASSERT_*` provides rich failure messages (`<< "fi_getinfo failed: " << fi_strerror(-ret)`), prints file/line, and integrates with test reporting. A shared C function would lose all of that.

3. Cleanup model differs: gtest relies on `TearDown()` always running. CMocka relies on registered teardown + `goto` skipping past failed setup. A shared helper would need to satisfy both models, making it more complex than two simple independent implementations.

The duplication is ~40 lines of straightforward sequential setup and teardown. The cost of deduplication (adapter layers, lost diagnostics) exceeds the cost of maintaining two copies.

Is this blocking? I am ready to push all the other changes requested.

**Comment by brianz98** [comment]
> Commits should be atomic based on what they do. gtest setup in the first commit, bugfix with its unit tests in the second commit.

The setup and the unit tests are logically one unit, all the setup is to serve the two added unit tests. I can separate them but it seems arbitrary to me. Do you want me to put everything in one commit instead if the bugfix and the unit test has to be in one commit?

**Comment by jiaxiyan** [comment]
> The setup and the unit tests are logically one unit, all the setup is to serve the two added unit tests. I can separate them but it seems arbitrary to me. Do you want me to put everything in one commit instead if the bugfix and the unit test has to be in one commit?

No, the first commit should be what is required for all unit tests. If we revert your first commit, your second commit will fail unit tests, which is not atomic.

File: prov/efa/test/gtest/efa_gtest_common_resource.cc (hunk 9 above)
Comment (jiaxiyan) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: If you insist this cannot be done, I won't block you.

Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: For the record, I never said it cannot be done, since it's a truly trivial change. I just don't think it's a change that will improve code quality for all the reasons I've given. Is there something you disagree with in there?

Comment (sunkuamzn) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: I have to say, I agree with Brian. I don't think there's a lot to gain by sharing code between the two frameworks.

File: prov/efa/test/gtest/efa_gtest_common_mocks.h (hunk 6 above)
Comment (sunkuamzn) [reply prov/efa/test/gtest/efa_gtest_common_mocks.h:49]: If a future test adds a new mocked function but the existing tests do not have an `EXPECT_CALL` for the new function, looks like the default gtest behavior is to print a warning but not fail the test. We could also use NiceMocks which wouldn't even print the warning.

Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_mocks.h:49]: Perhaps the worry would be that all mocked functions without expectation would return 0 / false / null or the default construct if a struct. If a real return different from these is needed, then we need to explicitly call the real function.

File: prov/efa/test/gtest/efa_gtest_common_resource.cc (hunk 9 above)
Comment (jiaxiyan) [reply prov/efa/test/gtest/efa_gtest_common_resource.cc:29]: efa_test_alloc_hints is the same as efa_unit_test_alloc_hints. We likely will need more functions from efa_unit_test_common.c as gtests expand. But like I said, not blocking.

**Comment by brianz98** [comment]
> > The setup and the unit tests are logically one unit, all the setup is to serve the two added unit tests. I can separate them but it seems arbitrary to me. Do you want me to put everything in one commit instead if the bugfix and the unit test has to be in one commit?
> 
> No, the first commit should be what is required for all unit tests. If we revert your first commit, your second commit will fail unit tests, which is not atomic.

Thanks for clarifying, done!

File: prov/efa/test/gtest/README.md (hunk 11 above)
Comment (brianz98) [reply prov/efa/test/gtest/README.md]: done!

File: prov/efa/test/gtest/README.md (hunk 10 above)
Comment (brianz98) [reply prov/efa/test/gtest/README.md]: done!

File: prov/efa/test/gtest/efa_gtest_common_helpers.h (hunk 7 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_helpers.h]: done!

File: prov/efa/test/gtest/efa_gtest_common_mocks.h (hunk 6 above)
Comment (brianz98) [reply prov/efa/test/gtest/efa_gtest_common_mocks.h:49]: Added instructions and a summary of below to README

> all mocked functions without expectation would return 0 / false / null or the default construct if a struct. If a real return different from these is needed, then we need to explicitly call the real function.

**Comment by brianz98** [comment]
Note: I have renamed `efa_gtest_common_device.{cc/h}` to `efa_gtest_common_mocks.{cc/h}`, to be more explicit (and correct) about what the files are doing. I have also rename the mock class from `EfaMockDevice` to `MockEfa`, and the corresponding instance name from `mock_device` to `mock_efa`. `Makefile.include` and `README.md` have been accordingly updated, latest changes compile and run fine.

**Review by sunkuamzn — DISMISSED** [review]

**Comment by a-szegel** [comment]
bot:aws:retest

**Review by a-szegel — APPROVED** [review]

**Review by sunkuamzn — APPROVED** [review]
